### ISSUE TITLE
Shape of noise action space of Pi0 (50x32) vs (32)

### ISSUE THREAD

**Issue by jdubie** [nakamotoo/dsrl_pi0#6]
@nakamotoo - great work - thanks for open sourcing the code!

The paper talks about a 1,600 (50x32) dimensional action space and it seems like the code only implements at 32 dimension action space. 1) Is this intentional? 2) If so are there alternatives you considered to upsampling 32 -> 50x32 other than using np.repeat? 3) Did you try pushing SAC to use a larger action space than 32?

From paper
> π0 uses an action chunk size of 50 and each action is 32-dimensional, resulting in a total action space dimension of 1,600

In code:
>         self.action_space = Box(low=-1, high=1, shape=(1, 32,), dtype=np.float32) # 32 is the noise action space of pi 0
https://github.com/nakamotoo/dsrl_pi0/blob/main/examples/train_real.py#L50

```py
                    # sac agent predicts the noise for diffusion model
                    actions_noise = agent.sample_actions(obs_dict)
                    actions_noise = np.reshape(actions_noise, agent.action_chunk_shape)
                    noise = np.repeat(actions_noise[-1:, :], 10 - actions_noise.shape[0], axis=0)
                    noise = jax.numpy.concatenate([actions_noise, noise], axis=0)[None]
```

**Comment by alexander-cobot** [comment]
I tried multiplying it by an interpolated version of itself along the time dimension

```python
noise = (
    einops.repeat(policy_out, "d -> n d", n=self._action_horizon)
    * jax.image.resize(policy_out, (self._action_horizon,), method=jax.image.ResizeMethod.LINEAR)[
        :, None
    ]
)
```

followed by a fixed-seed random shuffle

```python
noise = einops.rearrange(
    jax.random.permutation(jax.random.key(0), noise.flatten()),
    "(n d) -> n d",
    n=self._action_horizon,
)
```

For me, this helps with issues I was having. The robot was behaving erratically with the original paper's formulation - my guess is because the noise didn't look "noisy" enough.

**Comment by yachenkang** [comment]
> I tried multiplying it by an interpolated version of itself along the time dimension
> 
> noise = (
>     einops.repeat(policy_out, "d -> n d", n=self._action_horizon)
>     * jax.image.resize(policy_out, (self._action_horizon,), method=jax.image.ResizeMethod.LINEAR)[
>         :, None
>     ]
> )
> followed by a fixed-seed random shuffle
> 
> noise = einops.rearrange(
>     jax.random.permutation(jax.random.key(0), noise.flatten()),
>     "(n d) -> n d",
>     n=self._action_horizon,
> )
> For me, this helps with issues I was having. The robot was behaving erratically with the original paper's formulation - my guess is because the noise didn't look "noisy" enough.

I repeatedly experimented with this processing method. During the exploration phase, it indeed improved action stability and the task success rate. However, I suspect that this approach breaks the mapping that keeps the SAC policy output distribution and the DP policy input distribution on the same manifold. As a result, the SAC policy fails to learn useful information during training, which leads to the final task success rate not improving over the course of training. Do you have any successful experience in addressing this issue?

**Comment by zhoushaoxiang** [comment]
@alexander-cobot May I ask whether there are any plans to support DSRL in LeRobot?

**Comment by alexander-cobot** [comment]
Sorry for the lack of response. I didn't see these. I can't remember everything I did since I last posted this, but I know I settled on repeat + shuffle (no interpolate):

```python
obs["noise"] = einops.repeat(policy_out, "d -> n d", n=self._action_horizon)
obs["noise"] = einops.rearrange(
    jax.random.permutation(jax.random.key(0), obs["noise"].flatten()),
    "(n d) -> n d",
    n=self._action_horizon,
)
```

@yachenkang you say "I suspect that this approach breaks the mapping that keeps the SAC policy output distribution and the DP policy input distribution on the same manifold". Could have been because of the interpolation? Anyway I dropped that.

@zhoushaoxiang I have not been on the LeRobot team for a while now, so I'm not sure.

@pengzhi1998 @yachenkang the technique produced good results for me in terms of SR. Not sure if that started happening after I revised the method. But surely it had better exploration behaviour without the crazy jumps.

**Comment by pengzhi1998** [comment]
> Sorry for the lack of response. I didn't see these. I can't remember everything I did since I last posted this, but I know I settled on repeat + shuffle (no interpolate):
> 
> obs["noise"] = einops.repeat(policy_out, "d -> n d", n=self._action_horizon)
> obs["noise"] = einops.rearrange(
>     jax.random.permutation(jax.random.key(0), obs["noise"].flatten()),
>     "(n d) -> n d",
>     n=self._action_horizon,
> )
> [@yachenkang](https://github.com/yachenkang) you say "I suspect that this approach breaks the mapping that keeps the SAC policy output distribution and the DP policy input distribution on the same manifold". Could have been because of the interpolation? Anyway I dropped that.
> 
> [@zhoushaoxiang](https://github.com/zhoushaoxiang) I have not been on the LeRobot team for a while now, so I'm not sure.
> 
> [@pengzhi1998](https://github.com/pengzhi1998) [@yachenkang](https://github.com/yachenkang) the technique produced good results for me in terms of SR. Not sure if that started happening after I revised the method. But surely it had better exploration behaviour without the crazy jumps.

Dear Alexander,

Thank you for your reply!!!

That's great to know! I'm wondering whether you also increased the number of training steps before DSRL to improve the flow-matching head's robustness, so that it could deal with the repeated noise as well? (as discussed in several issues such as: https://github.com/nakamotoo/dsrl_pi0/issues/9 and https://github.com/nakamotoo/dsrl_pi0/issues/12, but the dilemma is: with increased training, the base policy using pure random noise would also perform better, which possibly makes DSRL not that necessary?)

Sorry for following up. Thank you again for your suggestions!

Best regards,
Pengzhi

**Comment by alexander-cobot** [comment]
@pengzhi1998 of course, I'm glad to share!

I didn't actually try DSRL for improving a policy trained for a target task. I tried it to select a desired behavior in a previously unseen env. So when I say "worked" it worked for that. That is to say, I can't answer your follow up question meaningfully because I was trying something very different.
